// @sendbird/uikit-react-native/lib/commonjs/domain/groupChannel/component/GroupChannelMessageList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeSafeAreaContext = require("react-native-safe-area-context");

var _uikitReactNativeFoundation = require("@sendbird/uikit-react-native-foundation");

var _uikitUtils = require("@sendbird/uikit-utils");

var _ChatFlatList = _interopRequireDefault(require("../../../components/ChatFlatList"));

var _ReactionAddons = require("../../../components/ReactionAddons");

var _constants = require("../../../constants");

var _useContext = require("../../../hooks/useContext");

var _SBUUtils = _interopRequireDefault(require("../../../libs/SBUUtils"));

var _moduleContext = require("../module/moduleContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const HANDLE_NEXT_MSG_SEPARATELY = _reactNative.Platform.select({
  default: true
});

const GroupChannelMessageList = _ref => {
  let {
    currentUserId,
    channel,
    messages,
    renderMessage,
    nextMessages,
    newMessagesFromMembers,
    onBottomReached,
    onTopReached,
    renderNewMessagesButton,
    renderScrollToBottomButton,
    onResendFailedMessage,
    onDeleteMessage,
    onPressImageMessage,
    onPressMediaMessage,
    flatListProps,
    enableMessageGrouping
  } = _ref;
  const {
    STRINGS
  } = (0, _useContext.useLocalization)();
  const {
    colors
  } = (0, _uikitReactNativeFoundation.useUIKitTheme)();
  const {
    left,
    right
  } = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const [scrollLeaveBottom, setScrollLeaveBottom] = (0, _react.useState)(false);
  const scrollRef = (0, _react.useRef)(null);
  const [newMessagesInternalBuffer, setNewMessagesInternalBuffer] = (0, _react.useState)(() => newMessagesFromMembers);
  const getMessagePressActions = useGetMessagePressActions({
    channel,
    currentUserId,
    onDeleteMessage,
    onResendFailedMessage,
    onPressImageMessage,
    onPressMediaMessage
  });
  const safeAreaLayout = {
    paddingLeft: left,
    paddingRight: right
  };
  const renderItem = (0, _uikitUtils.useFreshCallback)(_ref2 => {
    let {
      item,
      index
    } = _ref2;
    const {
      onPress,
      onLongPress
    } = getMessagePressActions(item);
    return renderMessage({
      message: item,
      prevMessage: messages[index + 1],
      nextMessage: messages[index - 1],
      onPress,
      onLongPress,
      enableMessageGrouping,
      channel,
      currentUserId
    });
  });

  if (!HANDLE_NEXT_MSG_SEPARATELY) {
    (0, _react.useEffect)(() => {
      if (newMessagesInternalBuffer.length !== 0) {
        setNewMessagesInternalBuffer(prev => prev.concat(newMessagesFromMembers));
      }

      onBottomReached();
    }, [newMessagesFromMembers]);
  }

  const onLeaveScrollBottom = (0, _react.useCallback)(val => {
    if (!HANDLE_NEXT_MSG_SEPARATELY) setNewMessagesInternalBuffer([]);
    setScrollLeaveBottom(val);
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      flex: 1,
      backgroundColor: colors.background
    }, safeAreaLayout]
  }, channel.isFrozen && /*#__PURE__*/_react.default.createElement(_uikitReactNativeFoundation.ChannelFrozenBanner, {
    style: styles.frozenBanner,
    text: STRINGS.GROUP_CHANNEL.LIST_BANNER_FROZEN
  }), /*#__PURE__*/_react.default.createElement(_ChatFlatList.default, _extends({
    nextMessages: nextMessages,
    onBottomReached: onBottomReached,
    onTopReached: onTopReached,
    onLeaveScrollBottom: onLeaveScrollBottom,
    currentUserId: currentUserId
  }, flatListProps, {
    listKey: `group-channel-messages-${channel.url}`,
    ref: scrollRef,
    data: messages,
    renderItem: renderItem,
    keyExtractor: _uikitUtils.messageKeyExtractor,
    contentContainerStyle: [// { minHeight: '100%', justifyContent: 'flex-end' },
    channel.isFrozen && styles.frozenListPadding, flatListProps === null || flatListProps === void 0 ? void 0 : flatListProps.contentContainerStyle]
  })), renderNewMessagesButton && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.newMsgButton, safeAreaLayout]
  }, renderNewMessagesButton({
    visible: scrollLeaveBottom,
    onPress: () => {
      var _scrollRef$current;

      return (_scrollRef$current = scrollRef.current) === null || _scrollRef$current === void 0 ? void 0 : _scrollRef$current.scrollToBottom(false);
    },
    newMessages: !HANDLE_NEXT_MSG_SEPARATELY ? newMessagesInternalBuffer : newMessagesFromMembers
  })), renderScrollToBottomButton && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    pointerEvents: scrollLeaveBottom ? 'auto' : 'none',
    style: [styles.scrollButton, safeAreaLayout]
  }, renderScrollToBottomButton({
    visible: scrollLeaveBottom,
    onPress: () => {
      var _scrollRef$current2;

      return (_scrollRef$current2 = scrollRef.current) === null || _scrollRef$current2 === void 0 ? void 0 : _scrollRef$current2.scrollToBottom(false);
    }
  })));
};

const useGetMessagePressActions = _ref3 => {
  let {
    channel,
    currentUserId,
    onResendFailedMessage,
    onDeleteMessage,
    onPressImageMessage,
    onPressMediaMessage
  } = _ref3;
  const {
    colors
  } = (0, _uikitReactNativeFoundation.useUIKitTheme)();
  const {
    STRINGS
  } = (0, _useContext.useLocalization)();
  const toast = (0, _uikitReactNativeFoundation.useToast)();
  const {
    openSheet
  } = (0, _uikitReactNativeFoundation.useBottomSheet)();
  const {
    alert
  } = (0, _uikitReactNativeFoundation.useAlert)();
  const {
    clipboardService,
    fileService
  } = (0, _useContext.usePlatformService)();
  const {
    features
  } = (0, _useContext.useSendbirdChat)();
  const {
    setMessageToEdit
  } = (0, _react.useContext)(_moduleContext.GroupChannelContexts.Fragment);

  const handleFailedMessage = message => {
    openSheet({
      sheetItems: [{
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_FAILED_RETRY,
        onPress: () => onResendFailedMessage(message).catch(() => toast.show(STRINGS.TOAST.RESEND_MSG_ERROR, 'error'))
      }, {
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_FAILED_REMOVE,
        titleColor: colors.ui.dialog.default.none.destructive,
        onPress: () => confirmDelete(message)
      }]
    });
  };

  const confirmDelete = message => {
    alert({
      title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_TITLE,
      buttons: [{
        text: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_CANCEL
      }, {
        text: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_OK,
        style: 'destructive',
        onPress: () => onDeleteMessage(message).catch(() => toast.show(STRINGS.TOAST.DELETE_MSG_ERROR, 'error'))
      }]
    });
  };

  return msg => {
    if (!msg.isUserMessage() && !msg.isFileMessage()) {
      return {
        onPress: undefined,
        onLongPress: undefined
      };
    }

    const sheetItems = [];
    const response = {
      onPress: undefined,
      onLongPress: undefined
    };

    if (msg.isUserMessage()) {
      sheetItems.push({
        icon: 'copy',
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_COPY,
        onPress: () => {
          clipboardService.setString(msg.message || '');
          toast.show(STRINGS.TOAST.COPY_OK, 'success');
        }
      });

      if ((0, _uikitUtils.isMyMessage)(msg, currentUserId) && msg.sendingStatus === 'succeeded') {
        sheetItems.push({
          icon: 'edit',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_EDIT,
          onPress: () => setMessageToEdit(msg)
        }, {
          icon: 'delete',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE,
          onPress: () => confirmDelete(msg)
        });
      }
    }

    if (msg.isFileMessage()) {
      sheetItems.push({
        icon: 'download',
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_SAVE,
        onPress: async () => {
          if ((0, _uikitUtils.toMegabyte)(msg.size) > 4) {
            toast.show(STRINGS.TOAST.DOWNLOAD_START, 'success');
          }

          fileService.save({
            fileUrl: msg.url,
            fileName: msg.name,
            fileType: msg.type
          }).then(response => {
            toast.show(STRINGS.TOAST.DOWNLOAD_OK, 'success');

            _uikitUtils.Logger.log('File saved to', response);
          }).catch(err => {
            toast.show(STRINGS.TOAST.DOWNLOAD_ERROR, 'error');

            _uikitUtils.Logger.log('File save failure', err);
          });
        }
      });

      if ((0, _uikitUtils.isMyMessage)(msg, currentUserId) && msg.sendingStatus === 'succeeded') {
        sheetItems.push({
          icon: 'delete',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE,
          onPress: () => confirmDelete(msg)
        });
      }

      const fileType = (0, _uikitUtils.getFileType)(msg.type || (0, _uikitUtils.getFileExtension)(msg.name));

      switch (fileType) {
        case 'image':
        case 'video':
        case 'audio':
          {
            response.onPress = () => {
              if (onPressImageMessage && fileType === 'image') {
                _uikitUtils.Logger.warn(_constants.DEPRECATION_WARNING.GROUP_CHANNEL.ON_PRESS_IMAGE_MESSAGE);

                onPressImageMessage(msg, (0, _uikitUtils.getAvailableUriFromFileMessage)(msg));
              }

              onPressMediaMessage === null || onPressMediaMessage === void 0 ? void 0 : onPressMediaMessage(msg, () => onDeleteMessage(msg), (0, _uikitUtils.getAvailableUriFromFileMessage)(msg));
            };

            break;
          }

        default:
          {
            response.onPress = () => _SBUUtils.default.openURL(msg.url);

            break;
          }
      }
    }

    if (sheetItems.length > 0) {
      response.onLongPress = () => {
        openSheet({
          sheetItems,
          HeaderComponent: (0, _uikitUtils.shouldRenderReaction)(channel, features.reactionEnabled) ? _ref4 => {
            let {
              onClose
            } = _ref4;
            return /*#__PURE__*/_react.default.createElement(_ReactionAddons.ReactionAddons.BottomSheet, {
              message: msg,
              channel: channel,
              onClose: onClose
            });
          } : undefined
        });
      };
    }

    if (msg.sendingStatus === 'failed') {
      response.onLongPress = () => handleFailedMessage(msg);

      response.onPress = () => {
        onResendFailedMessage(msg).catch(() => toast.show(STRINGS.TOAST.RESEND_MSG_ERROR, 'error'));
      };
    }

    if (msg.sendingStatus === 'pending') {
      response.onLongPress = undefined;
      response.onPress = undefined;
    }

    return response;
  };
};

const styles = (0, _uikitReactNativeFoundation.createStyleSheet)({
  frozenBanner: {
    position: 'absolute',
    zIndex: 999,
    top: 8,
    left: 8,
    right: 8
  },
  frozenListPadding: {
    paddingBottom: 32
  },
  newMsgButton: {
    position: 'absolute',
    zIndex: 999,
    bottom: 10,
    alignSelf: 'center'
  },
  scrollButton: {
    position: 'absolute',
    zIndex: 998,
    bottom: 10,
    right: 16
  }
});

var _default = /*#__PURE__*/_react.default.memo(GroupChannelMessageList);

exports.default = _default;
