// @sendbird/uikit-utils/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  NOOP: true,
  ASYNC_NOOP: true,
  PASS: true,
  toMegabyte: true,
  isFunction: true,
  SBErrorCode: true,
  SBErrorMessage: true,
  Logger: true,
  arrayToMap: true,
  arrayToMapWithGetter: true,
  createPubSub: true,
  urlRegexStrict: true,
  urlRegexRough: true,
  newLineRegex: true,
  audioExtRegex: true,
  videoExtRegex: true,
  imageExtRegex: true,
  getFileType: true,
  getFileExtension: true,
  replaceUrlAsComponents: true,
  normalizeFileName: true,
  emailRegex: true,
  phoneRegex: true,
  replaceWithRegex: true,
  createMentionTemplateRegex: true,
  BufferedRequest: true
};
exports.ASYNC_NOOP = void 0;
Object.defineProperty(exports, "BufferedRequest", {
  enumerable: true,
  get: function () {
    return _bufferedRequest.BufferedRequest;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _logger.Logger;
  }
});
exports.SBErrorMessage = exports.SBErrorCode = exports.PASS = exports.NOOP = void 0;
Object.defineProperty(exports, "arrayToMap", {
  enumerable: true,
  get: function () {
    return _arrayToMap.default;
  }
});
Object.defineProperty(exports, "arrayToMapWithGetter", {
  enumerable: true,
  get: function () {
    return _arrayToMap.arrayToMapWithGetter;
  }
});
Object.defineProperty(exports, "audioExtRegex", {
  enumerable: true,
  get: function () {
    return _regex.audioExtRegex;
  }
});
Object.defineProperty(exports, "createMentionTemplateRegex", {
  enumerable: true,
  get: function () {
    return _regex.createMentionTemplateRegex;
  }
});
Object.defineProperty(exports, "createPubSub", {
  enumerable: true,
  get: function () {
    return _pubsub.createPubSub;
  }
});
Object.defineProperty(exports, "emailRegex", {
  enumerable: true,
  get: function () {
    return _regex.emailRegex;
  }
});
Object.defineProperty(exports, "getFileExtension", {
  enumerable: true,
  get: function () {
    return _regex.getFileExtension;
  }
});
Object.defineProperty(exports, "getFileType", {
  enumerable: true,
  get: function () {
    return _regex.getFileType;
  }
});
Object.defineProperty(exports, "imageExtRegex", {
  enumerable: true,
  get: function () {
    return _regex.imageExtRegex;
  }
});
exports.isFunction = void 0;
Object.defineProperty(exports, "newLineRegex", {
  enumerable: true,
  get: function () {
    return _regex.newLineRegex;
  }
});
Object.defineProperty(exports, "normalizeFileName", {
  enumerable: true,
  get: function () {
    return _regex.normalizeFileName;
  }
});
Object.defineProperty(exports, "phoneRegex", {
  enumerable: true,
  get: function () {
    return _regex.phoneRegex;
  }
});
Object.defineProperty(exports, "replaceUrlAsComponents", {
  enumerable: true,
  get: function () {
    return _regex.replaceUrlAsComponents;
  }
});
Object.defineProperty(exports, "replaceWithRegex", {
  enumerable: true,
  get: function () {
    return _regex.replaceWithRegex;
  }
});
exports.toMegabyte = void 0;
Object.defineProperty(exports, "urlRegexRough", {
  enumerable: true,
  get: function () {
    return _regex.urlRegexRough;
  }
});
Object.defineProperty(exports, "urlRegexStrict", {
  enumerable: true,
  get: function () {
    return _regex.urlRegexStrict;
  }
});
Object.defineProperty(exports, "videoExtRegex", {
  enumerable: true,
  get: function () {
    return _regex.videoExtRegex;
  }
});

var _logger = require("./shared/logger");

var _arrayToMap = _interopRequireWildcard(require("./shared/arrayToMap"));

var _pubsub = require("./shared/pubsub");

var _regex = require("./shared/regex");

var _bufferedRequest = require("./shared/bufferedRequest");

var _shared = require("./shared");

Object.keys(_shared).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _shared[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _shared[key];
    }
  });
});

var _hooks = require("./hooks");

Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});

var _reactNative = require("./hooks/react-native");

Object.keys(_reactNative).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reactNative[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reactNative[key];
    }
  });
});

var _groupChannel = require("./ui-format/groupChannel");

Object.keys(_groupChannel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _groupChannel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _groupChannel[key];
    }
  });
});

var _common = require("./ui-format/common");

Object.keys(_common).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common[key];
    }
  });
});

var _channel = require("./sendbird/channel");

Object.keys(_channel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _channel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _channel[key];
    }
  });
});

var _message = require("./sendbird/message");

Object.keys(_message).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _message[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _message[key];
    }
  });
});

var _attrs = require("./sendbird/attrs");

Object.keys(_attrs).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _attrs[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _attrs[key];
    }
  });
});

var _user = require("./sendbird/user");

Object.keys(_user).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _user[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _user[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const NOOP = () => void 0;

exports.NOOP = NOOP;

const ASYNC_NOOP = async () => void 0;

exports.ASYNC_NOOP = ASYNC_NOOP;

const PASS = val => val;

exports.PASS = PASS;

const toMegabyte = byte => byte / 1024 / 1024;

exports.toMegabyte = toMegabyte;

const isFunction = param => typeof param === 'function';

exports.isFunction = isFunction;
const SBErrorCode = {
  NON_AUTHORIZED: 400108
};
exports.SBErrorCode = SBErrorCode;
const SBErrorMessage = {
  ACL: 'Sendbird provides various access control options when using the Chat SDK.\n' + 'By default, the Allow retrieving user list attribute is turned on to facilitate creating sample apps.\n' + 'However, this may grant access to unwanted data or operations, leading to potential security concerns.\n' + 'To manage your access control settings, you can turn on or off each setting on Sendbird Dashboard.'
};
exports.SBErrorMessage = SBErrorMessage;
