// @sendbird/uikit-react-native/lib/module/libs/SBUError.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SBUErrorCode;

(function (SBUErrorCode) {
  SBUErrorCode[SBUErrorCode["ERR_UNKNOWN"] = 90000000] = "ERR_UNKNOWN";
  SBUErrorCode[SBUErrorCode["ERR_PERMISSIONS_DENIED"] = 91001000] = "ERR_PERMISSIONS_DENIED";
  SBUErrorCode[SBUErrorCode["ERR_DEVICE_UNAVAILABLE"] = 91001001] = "ERR_DEVICE_UNAVAILABLE";
})(SBUErrorCode || (SBUErrorCode = {}));

export default class SBUError extends Error {
  static get UNKNOWN() {
    return new SBUError(SBUErrorCode.ERR_UNKNOWN);
  }

  static get PERMISSIONS_DENIED() {
    return new SBUError(SBUErrorCode.ERR_PERMISSIONS_DENIED);
  }

  static get DEVICE_UNAVAILABLE() {
    return new SBUError(SBUErrorCode.ERR_DEVICE_UNAVAILABLE);
  }

  constructor(code, message) {
    super(message);
    this.code = code;
  }

}

_defineProperty(SBUError, "CODE", SBUErrorCode);
