// @sendbird/uikit-react-native/lib/module/domain/groupChannel/component/GroupChannelInput/EditInput.js
import React, { forwardRef } from 'react';
import { Platform, View } from 'react-native';
import { MentionType } from '@sendbird/chat/message';
import { Button, TextInput, createStyleSheet, useToast } from '@sendbird/uikit-react-native-foundation';
import { useLocalization, useSendbirdChat } from '../../../../hooks/useContext';
const EditInput = /*#__PURE__*/forwardRef(function EditInput(_ref, ref) {
  let {
    text,
    onChangeText,
    messageToEdit,
    setMessageToEdit,
    onUpdateUserMessage,
    onSelectionChange,
    disabled,
    autoFocus,
    mentionedUsers
  } = _ref;
  const {
    mentionManager
  } = useSendbirdChat();
  const {
    STRINGS
  } = useLocalization();
  const toast = useToast();

  const onPressCancel = () => {
    setMessageToEdit();
    onChangeText('');
  };

  const onPressSave = () => {
    if (messageToEdit.isUserMessage()) {
      const mention = {
        userIds: mentionedUsers.map(it => it.user.userId),
        messageTemplate: mentionManager.textToMentionedMessageTemplate(text, mentionedUsers),
        type: MentionType.USERS
      };
      onUpdateUserMessage(text, messageToEdit, mention).catch(() => toast.show(STRINGS.TOAST.UPDATE_MSG_ERROR, 'error'));
    }

    setMessageToEdit();
    onChangeText('');
  };

  return /*#__PURE__*/React.createElement(View, {
    style: styles.editInputContainer
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.inputWrapper
  }, /*#__PURE__*/React.createElement(TextInput, {
    ref: ref,
    multiline: true,
    disableFullscreenUI: true,
    editable: !disabled,
    autoFocus: autoFocus,
    onChangeText: onChangeText,
    style: styles.input,
    placeholder: STRINGS.GROUP_CHANNEL.INPUT_PLACEHOLDER_ACTIVE,
    onSelectionChange: onSelectionChange
  }, mentionManager.textToMentionedComponents(text, mentionedUsers))), /*#__PURE__*/React.createElement(View, {
    style: {
      marginTop: 8,
      flexDirection: 'row'
    }
  }, /*#__PURE__*/React.createElement(Button, {
    variant: 'text',
    onPress: onPressCancel
  }, STRINGS.GROUP_CHANNEL.INPUT_EDIT_CANCEL), /*#__PURE__*/React.createElement(View, {
    style: styles.space
  }), /*#__PURE__*/React.createElement(Button, {
    variant: 'contained',
    onPress: onPressSave
  }, STRINGS.GROUP_CHANNEL.INPUT_EDIT_OK)));
});
const styles = createStyleSheet({
  editInputContainer: {
    paddingVertical: 10,
    paddingHorizontal: 16,
    flexDirection: 'column',
    alignItems: 'center'
  },
  input: {
    flex: 1,
    marginRight: 4,
    minHeight: 36,
    maxHeight: 36 * Platform.select({
      ios: 2.5,
      default: 2
    }),
    borderRadius: 20
  },
  inputWrapper: {
    flexDirection: 'row'
  },
  space: {
    flex: 1
  }
});
export default EditInput;
