// @sendbird/uikit-react-native/lib/module/domain/groupChannel/component/GroupChannelInput/index.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useContext, useEffect, useRef, useState } from 'react';
import { KeyboardAvoidingView, Platform, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { createStyleSheet, useUIKitTheme } from '@sendbird/uikit-react-native-foundation';
import { getGroupChannelChatAvailableState, replace, useIIFE } from '@sendbird/uikit-utils';
import { useSendbirdChat } from '../../../../hooks/useContext';
import useMentionTextInput from '../../../../hooks/useMentionTextInput';
import { GroupChannelContexts } from '../../module/moduleContext';
import EditInput from './EditInput';
import SendInput from './SendInput';
const AUTO_FOCUS = Platform.select({
  ios: false,
  android: true,
  default: false
});
const KEYBOARD_AVOID_VIEW_BEHAVIOR = Platform.select({
  ios: 'padding',
  default: undefined
}); // FIXME(iOS): Dynamic style does not work properly when typing the CJK. (https://github.com/facebook/react-native/issues/26107)
//  To workaround temporarily, change the key for re-mount the component.
//  -> This will affect to keyboard blur when add/remove first mentioned user.

const GET_INPUT_KEY = shouldReset => shouldReset ? 'uikit-input-clear' : 'uikit-input'; // TODO: Refactor 'Edit' mode to clearly


const GroupChannelInput = props => {
  const {
    top,
    left,
    right,
    bottom
  } = useSafeAreaInsets();
  const {
    colors
  } = useUIKitTheme();
  const {
    features,
    mentionManager
  } = useSendbirdChat();
  const {
    channel,
    messageToEdit,
    setMessageToEdit,
    keyboardAvoidOffset = 0
  } = useContext(GroupChannelContexts.Fragment);
  const chatAvailableState = getGroupChannelChatAvailableState(channel);
  const mentionAvailable = features.userMentionEnabled && channel.isGroupChannel() && !channel.isBroadcast;
  const inputMode = useIIFE(() => {
    if (!messageToEdit) return 'send';
    if (messageToEdit.isFileMessage()) return 'send';
    return 'edit';
  });
  const [inputHeight, setInputHeight] = useState(styles.inputDefault.height);
  const {
    selection,
    onSelectionChange,
    textInputRef,
    text,
    onChangeText,
    mentionedUsers
  } = useMentionTextInput({
    messageToEdit: messageToEdit
  });
  useTypingTrigger(text, channel);
  useTextPersistenceOnDisabled(text, onChangeText, chatAvailableState.disabled);
  useAutoFocusOnEditMode(textInputRef, messageToEdit);

  const onPressToMention = (user, searchStringRange) => {
    const mentionedMessageText = mentionManager.asMentionedMessageText(user, true);
    const range = {
      start: searchStringRange.start,
      end: searchStringRange.start + mentionedMessageText.length - 1
    };
    onChangeText(replace(text, searchStringRange.start, searchStringRange.end, mentionedMessageText), {
      user,
      range
    });
  };

  if (!props.shouldRenderInput) {
    return /*#__PURE__*/React.createElement(SafeAreaBottom, {
      height: bottom
    });
  }

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(KeyboardAvoidingView, {
    keyboardVerticalOffset: -bottom + keyboardAvoidOffset,
    behavior: KEYBOARD_AVOID_VIEW_BEHAVIOR
  }, /*#__PURE__*/React.createElement(View, {
    style: {
      paddingLeft: left,
      paddingRight: right,
      backgroundColor: colors.background
    }
  }, /*#__PURE__*/React.createElement(View, {
    onLayout: e => setInputHeight(e.nativeEvent.layout.height),
    style: styles.inputContainer
  }, inputMode === 'send' && /*#__PURE__*/React.createElement(SendInput, _extends({}, props, chatAvailableState, {
    key: GET_INPUT_KEY(mentionedUsers.length === 0),
    ref: textInputRef,
    text: text,
    onChangeText: onChangeText,
    onSelectionChange: onSelectionChange,
    mentionedUsers: mentionedUsers
  })), inputMode === 'edit' && messageToEdit && /*#__PURE__*/React.createElement(EditInput, _extends({}, props, {
    key: GET_INPUT_KEY(mentionedUsers.length === 0),
    ref: textInputRef,
    autoFocus: AUTO_FOCUS,
    text: text,
    onChangeText: onChangeText,
    messageToEdit: messageToEdit,
    setMessageToEdit: setMessageToEdit,
    disabled: chatAvailableState.disabled,
    onSelectionChange: onSelectionChange,
    mentionedUsers: mentionedUsers
  }))), /*#__PURE__*/React.createElement(SafeAreaBottom, {
    height: bottom
  }))), mentionAvailable && /*#__PURE__*/React.createElement(props.SuggestedMentionList, {
    text: text,
    selection: selection,
    inputHeight: inputHeight,
    topInset: top,
    bottomInset: bottom,
    onPressToMention: onPressToMention,
    mentionedUsers: mentionedUsers
  }));
};

const useTypingTrigger = (text, channel) => {
  useEffect(() => {
    if (text.length === 0) channel.endTyping();else channel.startTyping();
  }, [text]);
};

const useTextPersistenceOnDisabled = (text, setText, chatDisabled) => {
  const textTmpRef = useRef('');
  useEffect(() => {
    if (chatDisabled) {
      textTmpRef.current = text;
      setText('');
    } else {
      setText(textTmpRef.current);
    }
  }, [chatDisabled]);
};

const useAutoFocusOnEditMode = (textInputRef, messageToEdit) => {
  useEffect(() => {
    if (messageToEdit !== null && messageToEdit !== void 0 && messageToEdit.isUserMessage()) {
      if (!AUTO_FOCUS) setTimeout(() => {
        var _textInputRef$current;

        return (_textInputRef$current = textInputRef.current) === null || _textInputRef$current === void 0 ? void 0 : _textInputRef$current.focus();
      }, 500);
    }
  }, [messageToEdit]);
};

const SafeAreaBottom = _ref => {
  let {
    height
  } = _ref;
  return /*#__PURE__*/React.createElement(View, {
    style: {
      height
    }
  });
};

const styles = createStyleSheet({
  inputContainer: {
    justifyContent: 'center',
    width: '100%'
  },
  inputDefault: {
    height: 56
  }
});
export default /*#__PURE__*/React.memo(GroupChannelInput);
