// @sendbird/uikit-react-native/lib/commonjs/domain/groupChannelSettings/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GroupChannelSettingsContexts", {
  enumerable: true,
  get: function () {
    return _moduleContext.GroupChannelSettingsContexts;
  }
});
Object.defineProperty(exports, "GroupChannelSettingsContextsProvider", {
  enumerable: true,
  get: function () {
    return _moduleContext.GroupChannelSettingsContextsProvider;
  }
});
Object.defineProperty(exports, "GroupChannelSettingsHeader", {
  enumerable: true,
  get: function () {
    return _GroupChannelSettingsHeader.default;
  }
});
Object.defineProperty(exports, "GroupChannelSettingsInfo", {
  enumerable: true,
  get: function () {
    return _GroupChannelSettingsInfo.default;
  }
});
Object.defineProperty(exports, "GroupChannelSettingsMenu", {
  enumerable: true,
  get: function () {
    return _GroupChannelSettingsMenu.default;
  }
});
Object.defineProperty(exports, "createGroupChannelSettingsModule", {
  enumerable: true,
  get: function () {
    return _createGroupChannelSettingsModule.default;
  }
});

var _GroupChannelSettingsHeader = _interopRequireDefault(require("./component/GroupChannelSettingsHeader"));

var _GroupChannelSettingsInfo = _interopRequireDefault(require("./component/GroupChannelSettingsInfo"));

var _GroupChannelSettingsMenu = _interopRequireDefault(require("./component/GroupChannelSettingsMenu"));

var _createGroupChannelSettingsModule = _interopRequireDefault(require("./module/createGroupChannelSettingsModule"));

var _moduleContext = require("./module/moduleContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
