// @sendbird/uikit-react-native-foundation/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ActionMenu", {
  enumerable: true,
  get: function () {
    return _ActionMenu.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function () {
    return _Alert.default;
  }
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function () {
    return _Avatar.default;
  }
});
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function () {
    return _Badge.default;
  }
});
Object.defineProperty(exports, "BottomSheet", {
  enumerable: true,
  get: function () {
    return _BottomSheet.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function () {
    return _Button.default;
  }
});
Object.defineProperty(exports, "ChannelFrozenBanner", {
  enumerable: true,
  get: function () {
    return _ChannelFrozenBanner.default;
  }
});
Object.defineProperty(exports, "DarkUIKitTheme", {
  enumerable: true,
  get: function () {
    return _DarkUIKitTheme.default;
  }
});
Object.defineProperty(exports, "DialogProvider", {
  enumerable: true,
  get: function () {
    return _Dialog.DialogProvider;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _Divider.default;
  }
});
Object.defineProperty(exports, "GroupChannelPreview", {
  enumerable: true,
  get: function () {
    return _GroupChannelPreview.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function () {
    return _Header.default;
  }
});
Object.defineProperty(exports, "HeaderStyleContext", {
  enumerable: true,
  get: function () {
    return _HeaderStyleContext.HeaderStyleContext;
  }
});
Object.defineProperty(exports, "HeaderStyleProvider", {
  enumerable: true,
  get: function () {
    return _HeaderStyleContext.HeaderStyleProvider;
  }
});
Object.defineProperty(exports, "Icon", {
  enumerable: true,
  get: function () {
    return _Icon.default;
  }
});
Object.defineProperty(exports, "IconAssets", {
  enumerable: true,
  get: function () {
    return _icon.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _Image.default;
  }
});
Object.defineProperty(exports, "LightUIKitTheme", {
  enumerable: true,
  get: function () {
    return _LightUIKitTheme.default;
  }
});
Object.defineProperty(exports, "LoadingSpinner", {
  enumerable: true,
  get: function () {
    return _LoadingSpinner.default;
  }
});
Object.defineProperty(exports, "MenuBar", {
  enumerable: true,
  get: function () {
    return _MenuBar.default;
  }
});
Object.defineProperty(exports, "MenuBarProps", {
  enumerable: true,
  get: function () {
    return _MenuBar.MenuBarProps;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function () {
    return _Modal.default;
  }
});
Object.defineProperty(exports, "OutlinedButton", {
  enumerable: true,
  get: function () {
    return _OutlinedButton.default;
  }
});
Object.defineProperty(exports, "Palette", {
  enumerable: true,
  get: function () {
    return _Palette.default;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function () {
    return _Placeholder.default;
  }
});
Object.defineProperty(exports, "ProfileCard", {
  enumerable: true,
  get: function () {
    return _ProfileCard.default;
  }
});
Object.defineProperty(exports, "Prompt", {
  enumerable: true,
  get: function () {
    return _Prompt.default;
  }
});
Object.defineProperty(exports, "RegexText", {
  enumerable: true,
  get: function () {
    return _RegexText.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function () {
    return _Switch.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _Text.default;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function () {
    return _TextInput.default;
  }
});
Object.defineProperty(exports, "Toast", {
  enumerable: true,
  get: function () {
    return _Toast.default;
  }
});
Object.defineProperty(exports, "ToastProvider", {
  enumerable: true,
  get: function () {
    return _Toast.ToastProvider;
  }
});
Object.defineProperty(exports, "UIKitThemeContext", {
  enumerable: true,
  get: function () {
    return _UIKitThemeContext.default;
  }
});
Object.defineProperty(exports, "UIKitThemeProvider", {
  enumerable: true,
  get: function () {
    return _UIKitThemeProvider.default;
  }
});
Object.defineProperty(exports, "URLParsedText", {
  enumerable: true,
  get: function () {
    return _URLParsedText.default;
  }
});
Object.defineProperty(exports, "createScaleFactor", {
  enumerable: true,
  get: function () {
    return _createScaleFactor.default;
  }
});
Object.defineProperty(exports, "createSelectByColorScheme", {
  enumerable: true,
  get: function () {
    return _createSelectByColorScheme.default;
  }
});
Object.defineProperty(exports, "createStyleSheet", {
  enumerable: true,
  get: function () {
    return _createStyleSheet.default;
  }
});
Object.defineProperty(exports, "createTheme", {
  enumerable: true,
  get: function () {
    return _createTheme.default;
  }
});
Object.defineProperty(exports, "createTypography", {
  enumerable: true,
  get: function () {
    return _createTypography.default;
  }
});
Object.defineProperty(exports, "getDefaultHeaderHeight", {
  enumerable: true,
  get: function () {
    return _getDefaultHeaderHeight.default;
  }
});
Object.defineProperty(exports, "useActionMenu", {
  enumerable: true,
  get: function () {
    return _Dialog.useActionMenu;
  }
});
Object.defineProperty(exports, "useAlert", {
  enumerable: true,
  get: function () {
    return _Dialog.useAlert;
  }
});
Object.defineProperty(exports, "useBottomSheet", {
  enumerable: true,
  get: function () {
    return _Dialog.useBottomSheet;
  }
});
Object.defineProperty(exports, "useHeaderStyle", {
  enumerable: true,
  get: function () {
    return _useHeaderStyle.default;
  }
});
Object.defineProperty(exports, "usePrompt", {
  enumerable: true,
  get: function () {
    return _Dialog.usePrompt;
  }
});
Object.defineProperty(exports, "useToast", {
  enumerable: true,
  get: function () {
    return _Toast.useToast;
  }
});
Object.defineProperty(exports, "useUIKitTheme", {
  enumerable: true,
  get: function () {
    return _useUIKitTheme.default;
  }
});

var _icon = _interopRequireDefault(require("./assets/icon"));

var _Divider = _interopRequireDefault(require("./components/Divider"));

var _Icon = _interopRequireDefault(require("./components/Icon"));

var _Image = _interopRequireDefault(require("./components/Image"));

var _Modal = _interopRequireDefault(require("./components/Modal"));

var _RegexText = _interopRequireDefault(require("./components/RegexText"));

var _Switch = _interopRequireDefault(require("./components/Switch"));

var _Text = _interopRequireDefault(require("./components/Text"));

var _TextInput = _interopRequireDefault(require("./components/TextInput"));

var _URLParsedText = _interopRequireDefault(require("./components/URLParsedText"));

var _ActionMenu = _interopRequireDefault(require("./ui/ActionMenu"));

var _Alert = _interopRequireDefault(require("./ui/Alert"));

var _Avatar = _interopRequireDefault(require("./ui/Avatar"));

var _Badge = _interopRequireDefault(require("./ui/Badge"));

var _BottomSheet = _interopRequireDefault(require("./ui/BottomSheet"));

var _Button = _interopRequireDefault(require("./ui/Button"));

var _ChannelFrozenBanner = _interopRequireDefault(require("./ui/ChannelFrozenBanner"));

var _Dialog = require("./ui/Dialog");

var _GroupChannelPreview = _interopRequireDefault(require("./ui/GroupChannelPreview"));

var _Header = _interopRequireDefault(require("./ui/Header"));

var _LoadingSpinner = _interopRequireDefault(require("./ui/LoadingSpinner"));

var _MenuBar = _interopRequireWildcard(require("./ui/MenuBar"));

var _OutlinedButton = _interopRequireDefault(require("./ui/OutlinedButton"));

var _Placeholder = _interopRequireDefault(require("./ui/Placeholder"));

var _ProfileCard = _interopRequireDefault(require("./ui/ProfileCard"));

var _Prompt = _interopRequireDefault(require("./ui/Prompt"));

var _Toast = _interopRequireWildcard(require("./ui/Toast"));

var _createSelectByColorScheme = _interopRequireDefault(require("./styles/createSelectByColorScheme"));

var _createScaleFactor = _interopRequireDefault(require("./styles/createScaleFactor"));

var _createStyleSheet = _interopRequireDefault(require("./styles/createStyleSheet"));

var _createTypography = _interopRequireDefault(require("./styles/createTypography"));

var _getDefaultHeaderHeight = _interopRequireDefault(require("./styles/getDefaultHeaderHeight"));

var _HeaderStyleContext = require("./styles/HeaderStyleContext");

var _useHeaderStyle = _interopRequireDefault(require("./styles/useHeaderStyle"));

var _createTheme = _interopRequireDefault(require("./theme/createTheme"));

var _DarkUIKitTheme = _interopRequireDefault(require("./theme/DarkUIKitTheme"));

var _LightUIKitTheme = _interopRequireDefault(require("./theme/LightUIKitTheme"));

var _Palette = _interopRequireDefault(require("./theme/Palette"));

var _UIKitThemeContext = _interopRequireDefault(require("./theme/UIKitThemeContext"));

var _UIKitThemeProvider = _interopRequireDefault(require("./theme/UIKitThemeProvider"));

var _useUIKitTheme = _interopRequireDefault(require("./theme/useUIKitTheme"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
