// @sendbird/uikit-react-native/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  MessageRenderer: true,
  ReactionAddons: true,
  ReactionBottomSheets: true,
  ChannelCover: true,
  ChatFlatList: true,
  FileViewer: true,
  NewMessagesButton: true,
  ProviderLayout: true,
  ScrollToBottomButton: true,
  StatusComposition: true,
  TypedPlaceholder: true,
  UserActionBar: true,
  UserSelectableBar: true,
  createGroupChannelCreateFragment: true,
  createGroupChannelFragment: true,
  createGroupChannelSettingsFragment: true,
  createGroupChannelInviteFragment: true,
  createGroupChannelListFragment: true,
  createGroupChannelMembersFragment: true,
  createGroupChannelModerationFragment: true,
  createGroupChannelOperatorsFragment: true,
  createGroupChannelRegisterOperatorFragment: true,
  createGroupChannelMutedMembersFragment: true,
  createGroupChannelBannedUsersFragment: true,
  createGroupChannelNotificationsFragment: true,
  SendbirdChatContext: true,
  SendbirdChatProvider: true,
  PlatformServiceContext: true,
  PlatformServiceProvider: true,
  UserProfileContext: true,
  UserProfileProvider: true,
  LocalizationContext: true,
  LocalizationProvider: true,
  useConnection: true,
  usePushTokenRegistration: true,
  StringSetEn: true,
  createBaseStringSet: true,
  createNativeFileService: true,
  createNativeClipboardService: true,
  createNativeNotificationService: true,
  createNativeMediaService: true,
  createExpoFileService: true,
  createExpoClipboardService: true,
  createExpoNotificationService: true,
  createExpoMediaService: true,
  GroupChannelNotificationsProps: true,
  GroupChannelNotificationsContextsType: true,
  GroupChannelNotificationsFragment: true,
  GroupChannelNotificationsModule: true,
  SendbirdUIKitContainer: true,
  SendbirdUIKit: true,
  SBUError: true
};
Object.defineProperty(exports, "ChannelCover", {
  enumerable: true,
  get: function () {
    return _ChannelCover.default;
  }
});
Object.defineProperty(exports, "ChatFlatList", {
  enumerable: true,
  get: function () {
    return _ChatFlatList.default;
  }
});
Object.defineProperty(exports, "FileViewer", {
  enumerable: true,
  get: function () {
    return _FileViewer.default;
  }
});
Object.defineProperty(exports, "GroupChannelNotificationsContextsType", {
  enumerable: true,
  get: function () {
    return _types.GroupChannelNotificationsContextsType;
  }
});
Object.defineProperty(exports, "GroupChannelNotificationsFragment", {
  enumerable: true,
  get: function () {
    return _types.GroupChannelNotificationsFragment;
  }
});
Object.defineProperty(exports, "GroupChannelNotificationsModule", {
  enumerable: true,
  get: function () {
    return _types.GroupChannelNotificationsModule;
  }
});
Object.defineProperty(exports, "GroupChannelNotificationsProps", {
  enumerable: true,
  get: function () {
    return _types.GroupChannelNotificationsProps;
  }
});
Object.defineProperty(exports, "LocalizationContext", {
  enumerable: true,
  get: function () {
    return _LocalizationCtx.LocalizationContext;
  }
});
Object.defineProperty(exports, "LocalizationProvider", {
  enumerable: true,
  get: function () {
    return _LocalizationCtx.LocalizationProvider;
  }
});
Object.defineProperty(exports, "MessageRenderer", {
  enumerable: true,
  get: function () {
    return _MessageRenderer.default;
  }
});
Object.defineProperty(exports, "NewMessagesButton", {
  enumerable: true,
  get: function () {
    return _NewMessagesButton.default;
  }
});
Object.defineProperty(exports, "PlatformServiceContext", {
  enumerable: true,
  get: function () {
    return _PlatformServiceCtx.PlatformServiceContext;
  }
});
Object.defineProperty(exports, "PlatformServiceProvider", {
  enumerable: true,
  get: function () {
    return _PlatformServiceCtx.PlatformServiceProvider;
  }
});
Object.defineProperty(exports, "ProviderLayout", {
  enumerable: true,
  get: function () {
    return _ProviderLayout.default;
  }
});
Object.defineProperty(exports, "ReactionAddons", {
  enumerable: true,
  get: function () {
    return _ReactionAddons.ReactionAddons;
  }
});
Object.defineProperty(exports, "ReactionBottomSheets", {
  enumerable: true,
  get: function () {
    return _ReactionBottomSheets.ReactionBottomSheets;
  }
});
Object.defineProperty(exports, "SBUError", {
  enumerable: true,
  get: function () {
    return _SBUError.default;
  }
});
Object.defineProperty(exports, "ScrollToBottomButton", {
  enumerable: true,
  get: function () {
    return _ScrollToBottomButton.default;
  }
});
Object.defineProperty(exports, "SendbirdChatContext", {
  enumerable: true,
  get: function () {
    return _SendbirdChatCtx.SendbirdChatContext;
  }
});
Object.defineProperty(exports, "SendbirdChatProvider", {
  enumerable: true,
  get: function () {
    return _SendbirdChatCtx.SendbirdChatProvider;
  }
});
Object.defineProperty(exports, "SendbirdUIKit", {
  enumerable: true,
  get: function () {
    return _SendbirdUIKitContainer.SendbirdUIKit;
  }
});
Object.defineProperty(exports, "SendbirdUIKitContainer", {
  enumerable: true,
  get: function () {
    return _SendbirdUIKitContainer.default;
  }
});
Object.defineProperty(exports, "StatusComposition", {
  enumerable: true,
  get: function () {
    return _StatusComposition.default;
  }
});
Object.defineProperty(exports, "StringSetEn", {
  enumerable: true,
  get: function () {
    return _StringSet.default;
  }
});
Object.defineProperty(exports, "TypedPlaceholder", {
  enumerable: true,
  get: function () {
    return _TypedPlaceholder.default;
  }
});
Object.defineProperty(exports, "UserActionBar", {
  enumerable: true,
  get: function () {
    return _UserActionBar.default;
  }
});
Object.defineProperty(exports, "UserProfileContext", {
  enumerable: true,
  get: function () {
    return _UserProfileCtx.UserProfileContext;
  }
});
Object.defineProperty(exports, "UserProfileProvider", {
  enumerable: true,
  get: function () {
    return _UserProfileCtx.UserProfileProvider;
  }
});
Object.defineProperty(exports, "UserSelectableBar", {
  enumerable: true,
  get: function () {
    return _UserSelectableBar.default;
  }
});
Object.defineProperty(exports, "createBaseStringSet", {
  enumerable: true,
  get: function () {
    return _StringSet2.createBaseStringSet;
  }
});
Object.defineProperty(exports, "createExpoClipboardService", {
  enumerable: true,
  get: function () {
    return _createClipboardService2.default;
  }
});
Object.defineProperty(exports, "createExpoFileService", {
  enumerable: true,
  get: function () {
    return _createFileService2.default;
  }
});
Object.defineProperty(exports, "createExpoMediaService", {
  enumerable: true,
  get: function () {
    return _createMediaService2.default;
  }
});
Object.defineProperty(exports, "createExpoNotificationService", {
  enumerable: true,
  get: function () {
    return _createNotificationService2.default;
  }
});
Object.defineProperty(exports, "createGroupChannelBannedUsersFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelBannedUsersFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelCreateFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelCreateFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelInviteFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelInviteFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelListFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelListFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelMembersFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelMembersFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelModerationFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelModerationFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelMutedMembersFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelMutedMembersFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelNotificationsFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelNotificationsFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelOperatorsFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelOperatorsFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelRegisterOperatorFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelRegisterOperatorFragment.default;
  }
});
Object.defineProperty(exports, "createGroupChannelSettingsFragment", {
  enumerable: true,
  get: function () {
    return _createGroupChannelSettingsFragment.default;
  }
});
Object.defineProperty(exports, "createNativeClipboardService", {
  enumerable: true,
  get: function () {
    return _createClipboardService.default;
  }
});
Object.defineProperty(exports, "createNativeFileService", {
  enumerable: true,
  get: function () {
    return _createFileService.default;
  }
});
Object.defineProperty(exports, "createNativeMediaService", {
  enumerable: true,
  get: function () {
    return _createMediaService.default;
  }
});
Object.defineProperty(exports, "createNativeNotificationService", {
  enumerable: true,
  get: function () {
    return _createNotificationService.default;
  }
});
Object.defineProperty(exports, "useConnection", {
  enumerable: true,
  get: function () {
    return _useConnection.default;
  }
});
Object.defineProperty(exports, "usePushTokenRegistration", {
  enumerable: true,
  get: function () {
    return _usePushTokenRegistration.default;
  }
});

var _reactNative = require("react-native");

var _uikitUtils = require("@sendbird/uikit-utils");

var _MessageRenderer = _interopRequireDefault(require("./components/MessageRenderer"));

var _ReactionAddons = require("./components/ReactionAddons");

var _ReactionBottomSheets = require("./components/ReactionBottomSheets");

var _ChannelCover = _interopRequireDefault(require("./components/ChannelCover"));

var _ChatFlatList = _interopRequireDefault(require("./components/ChatFlatList"));

var _FileViewer = _interopRequireDefault(require("./components/FileViewer"));

var _NewMessagesButton = _interopRequireDefault(require("./components/NewMessagesButton"));

var _ProviderLayout = _interopRequireDefault(require("./components/ProviderLayout"));

var _ScrollToBottomButton = _interopRequireDefault(require("./components/ScrollToBottomButton"));

var _StatusComposition = _interopRequireDefault(require("./components/StatusComposition"));

var _TypedPlaceholder = _interopRequireDefault(require("./components/TypedPlaceholder"));

var _UserActionBar = _interopRequireDefault(require("./components/UserActionBar"));

var _UserSelectableBar = _interopRequireDefault(require("./components/UserSelectableBar"));

var _createGroupChannelCreateFragment = _interopRequireDefault(require("./fragments/createGroupChannelCreateFragment"));

var _createGroupChannelFragment = _interopRequireDefault(require("./fragments/createGroupChannelFragment"));

var _createGroupChannelSettingsFragment = _interopRequireDefault(require("./fragments/createGroupChannelSettingsFragment"));

var _createGroupChannelInviteFragment = _interopRequireDefault(require("./fragments/createGroupChannelInviteFragment"));

var _createGroupChannelListFragment = _interopRequireDefault(require("./fragments/createGroupChannelListFragment"));

var _createGroupChannelMembersFragment = _interopRequireDefault(require("./fragments/createGroupChannelMembersFragment"));

var _createGroupChannelModerationFragment = _interopRequireDefault(require("./fragments/createGroupChannelModerationFragment"));

var _createGroupChannelOperatorsFragment = _interopRequireDefault(require("./fragments/createGroupChannelOperatorsFragment"));

var _createGroupChannelRegisterOperatorFragment = _interopRequireDefault(require("./fragments/createGroupChannelRegisterOperatorFragment"));

var _createGroupChannelMutedMembersFragment = _interopRequireDefault(require("./fragments/createGroupChannelMutedMembersFragment"));

var _createGroupChannelBannedUsersFragment = _interopRequireDefault(require("./fragments/createGroupChannelBannedUsersFragment"));

var _createGroupChannelNotificationsFragment = _interopRequireDefault(require("./fragments/createGroupChannelNotificationsFragment"));

var _SendbirdChatCtx = require("./contexts/SendbirdChatCtx");

var _PlatformServiceCtx = require("./contexts/PlatformServiceCtx");

var _UserProfileCtx = require("./contexts/UserProfileCtx");

var _LocalizationCtx = require("./contexts/LocalizationCtx");

var _useConnection = _interopRequireDefault(require("./hooks/useConnection"));

var _usePushTokenRegistration = _interopRequireDefault(require("./hooks/usePushTokenRegistration"));

var _useContext = require("./hooks/useContext");

Object.keys(_useContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useContext[key];
    }
  });
});

var _StringSet = _interopRequireDefault(require("./localization/StringSet.en"));

var _StringSet2 = require("./localization/StringSet.type");

var _createFileService = _interopRequireDefault(require("./platform/createFileService.native"));

var _createClipboardService = _interopRequireDefault(require("./platform/createClipboardService.native"));

var _createNotificationService = _interopRequireDefault(require("./platform/createNotificationService.native"));

var _createMediaService = _interopRequireDefault(require("./platform/createMediaService.native"));

var _createFileService2 = _interopRequireDefault(require("./platform/createFileService.expo"));

var _createClipboardService2 = _interopRequireDefault(require("./platform/createClipboardService.expo"));

var _createNotificationService2 = _interopRequireDefault(require("./platform/createNotificationService.expo"));

var _createMediaService2 = _interopRequireDefault(require("./platform/createMediaService.expo"));

var _groupChannel = require("./domain/groupChannel");

Object.keys(_groupChannel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _groupChannel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _groupChannel[key];
    }
  });
});

var _groupChannelSettings = require("./domain/groupChannelSettings");

Object.keys(_groupChannelSettings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _groupChannelSettings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _groupChannelSettings[key];
    }
  });
});

var _groupChannelList = require("./domain/groupChannelList");

Object.keys(_groupChannelList).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _groupChannelList[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _groupChannelList[key];
    }
  });
});

var _groupChannelNotifications = require("./domain/groupChannelNotifications");

Object.keys(_groupChannelNotifications).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _groupChannelNotifications[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _groupChannelNotifications[key];
    }
  });
});

var _types = require("./domain/groupChannelNotifications/types");

var _types2 = require("./domain/groupChannelUserList/types");

Object.keys(_types2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types2[key];
    }
  });
});

var _userList = require("./domain/userList");

Object.keys(_userList).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _userList[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _userList[key];
    }
  });
});

var _SendbirdUIKitContainer = _interopRequireWildcard(require("./containers/SendbirdUIKitContainer"));

var _SBUError = _interopRequireDefault(require("./libs/SBUError"));

var _types3 = require("./types");

Object.keys(_types3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types3[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Components **/

/** Fragments **/

/** Context **/

/** Hooks **/

/** Localization **/

/** Platform API **/

/** Domain **/

/** UIKit **/
_uikitUtils.Logger.setLogLevel(__DEV__ ? 'warn' : 'none');

_uikitUtils.Logger.setTitle(`[UIKIT_${_reactNative.Platform.OS}]`);
