// @sendbird/uikit-react-native/lib/module/domain/groupChannel/component/GroupChannelMessageList.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { Platform, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { ChannelFrozenBanner, createStyleSheet, useAlert, useBottomSheet, useToast, useUIKitTheme } from '@sendbird/uikit-react-native-foundation';
import { Logger, getAvailableUriFromFileMessage, getFileExtension, getFileType, isMyMessage, messageKeyExtractor, shouldRenderReaction, toMegabyte, useFreshCallback } from '@sendbird/uikit-utils';
import ChatFlatList from '../../../components/ChatFlatList';
import { ReactionAddons } from '../../../components/ReactionAddons';
import { DEPRECATION_WARNING } from '../../../constants';
import { useLocalization, usePlatformService, useSendbirdChat } from '../../../hooks/useContext';
import SBUUtils from '../../../libs/SBUUtils';
import { GroupChannelContexts } from '../module/moduleContext';
const HANDLE_NEXT_MSG_SEPARATELY = Platform.select({
  default: true
});

const GroupChannelMessageList = _ref => {
  let {
    currentUserId,
    channel,
    messages,
    renderMessage,
    nextMessages,
    newMessagesFromMembers,
    onBottomReached,
    onTopReached,
    renderNewMessagesButton,
    renderScrollToBottomButton,
    onResendFailedMessage,
    onDeleteMessage,
    onPressImageMessage,
    onPressMediaMessage,
    flatListProps,
    enableMessageGrouping
  } = _ref;
  const {
    STRINGS
  } = useLocalization();
  const {
    colors
  } = useUIKitTheme();
  const {
    left,
    right
  } = useSafeAreaInsets();
  const [scrollLeaveBottom, setScrollLeaveBottom] = useState(false);
  const scrollRef = useRef(null);
  const [newMessagesInternalBuffer, setNewMessagesInternalBuffer] = useState(() => newMessagesFromMembers);
  const getMessagePressActions = useGetMessagePressActions({
    channel,
    currentUserId,
    onDeleteMessage,
    onResendFailedMessage,
    onPressImageMessage,
    onPressMediaMessage
  });
  const safeAreaLayout = {
    paddingLeft: left,
    paddingRight: right
  };
  const renderItem = useFreshCallback(_ref2 => {
    let {
      item,
      index
    } = _ref2;
    const {
      onPress,
      onLongPress
    } = getMessagePressActions(item);
    return renderMessage({
      message: item,
      prevMessage: messages[index + 1],
      nextMessage: messages[index - 1],
      onPress,
      onLongPress,
      enableMessageGrouping,
      channel,
      currentUserId
    });
  });

  if (!HANDLE_NEXT_MSG_SEPARATELY) {
    useEffect(() => {
      if (newMessagesInternalBuffer.length !== 0) {
        setNewMessagesInternalBuffer(prev => prev.concat(newMessagesFromMembers));
      }

      onBottomReached();
    }, [newMessagesFromMembers]);
  }

  const onLeaveScrollBottom = useCallback(val => {
    if (!HANDLE_NEXT_MSG_SEPARATELY) setNewMessagesInternalBuffer([]);
    setScrollLeaveBottom(val);
  }, []);
  return /*#__PURE__*/React.createElement(View, {
    style: [{
      flex: 1,
      backgroundColor: colors.background
    }, safeAreaLayout]
  }, channel.isFrozen && /*#__PURE__*/React.createElement(ChannelFrozenBanner, {
    style: styles.frozenBanner,
    text: STRINGS.GROUP_CHANNEL.LIST_BANNER_FROZEN
  }), /*#__PURE__*/React.createElement(ChatFlatList, _extends({
    nextMessages: nextMessages,
    onBottomReached: onBottomReached,
    onTopReached: onTopReached,
    onLeaveScrollBottom: onLeaveScrollBottom,
    currentUserId: currentUserId
  }, flatListProps, {
    listKey: `group-channel-messages-${channel.url}`,
    ref: scrollRef,
    data: messages,
    renderItem: renderItem,
    keyExtractor: messageKeyExtractor,
    contentContainerStyle: [// { minHeight: '100%', justifyContent: 'flex-end' },
    channel.isFrozen && styles.frozenListPadding, flatListProps === null || flatListProps === void 0 ? void 0 : flatListProps.contentContainerStyle]
  })), renderNewMessagesButton && /*#__PURE__*/React.createElement(View, {
    style: [styles.newMsgButton, safeAreaLayout]
  }, renderNewMessagesButton({
    visible: scrollLeaveBottom,
    onPress: () => {
      var _scrollRef$current;

      return (_scrollRef$current = scrollRef.current) === null || _scrollRef$current === void 0 ? void 0 : _scrollRef$current.scrollToBottom(false);
    },
    newMessages: !HANDLE_NEXT_MSG_SEPARATELY ? newMessagesInternalBuffer : newMessagesFromMembers
  })), renderScrollToBottomButton && /*#__PURE__*/React.createElement(View, {
    pointerEvents: scrollLeaveBottom ? 'auto' : 'none',
    style: [styles.scrollButton, safeAreaLayout]
  }, renderScrollToBottomButton({
    visible: scrollLeaveBottom,
    onPress: () => {
      var _scrollRef$current2;

      return (_scrollRef$current2 = scrollRef.current) === null || _scrollRef$current2 === void 0 ? void 0 : _scrollRef$current2.scrollToBottom(false);
    }
  })));
};

const useGetMessagePressActions = _ref3 => {
  let {
    channel,
    currentUserId,
    onResendFailedMessage,
    onDeleteMessage,
    onPressImageMessage,
    onPressMediaMessage
  } = _ref3;
  const {
    colors
  } = useUIKitTheme();
  const {
    STRINGS
  } = useLocalization();
  const toast = useToast();
  const {
    openSheet
  } = useBottomSheet();
  const {
    alert
  } = useAlert();
  const {
    clipboardService,
    fileService
  } = usePlatformService();
  const {
    features
  } = useSendbirdChat();
  const {
    setMessageToEdit
  } = useContext(GroupChannelContexts.Fragment);

  const handleFailedMessage = message => {
    openSheet({
      sheetItems: [{
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_FAILED_RETRY,
        onPress: () => onResendFailedMessage(message).catch(() => toast.show(STRINGS.TOAST.RESEND_MSG_ERROR, 'error'))
      }, {
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_FAILED_REMOVE,
        titleColor: colors.ui.dialog.default.none.destructive,
        onPress: () => confirmDelete(message)
      }]
    });
  };

  const confirmDelete = message => {
    alert({
      title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_TITLE,
      buttons: [{
        text: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_CANCEL
      }, {
        text: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE_CONFIRM_OK,
        style: 'destructive',
        onPress: () => onDeleteMessage(message).catch(() => toast.show(STRINGS.TOAST.DELETE_MSG_ERROR, 'error'))
      }]
    });
  };

  return msg => {
    if (!msg.isUserMessage() && !msg.isFileMessage()) {
      return {
        onPress: undefined,
        onLongPress: undefined
      };
    }

    const sheetItems = [];
    const response = {
      onPress: undefined,
      onLongPress: undefined
    };

    if (msg.isUserMessage()) {
      sheetItems.push({
        icon: 'copy',
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_COPY,
        onPress: () => {
          clipboardService.setString(msg.message || '');
          toast.show(STRINGS.TOAST.COPY_OK, 'success');
        }
      });

      if (isMyMessage(msg, currentUserId) && msg.sendingStatus === 'succeeded') {
        sheetItems.push({
          icon: 'edit',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_EDIT,
          onPress: () => setMessageToEdit(msg)
        }, {
          icon: 'delete',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE,
          onPress: () => confirmDelete(msg)
        });
      }
    }

    if (msg.isFileMessage()) {
      sheetItems.push({
        icon: 'download',
        title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_SAVE,
        onPress: async () => {
          if (toMegabyte(msg.size) > 4) {
            toast.show(STRINGS.TOAST.DOWNLOAD_START, 'success');
          }

          fileService.save({
            fileUrl: msg.url,
            fileName: msg.name,
            fileType: msg.type
          }).then(response => {
            toast.show(STRINGS.TOAST.DOWNLOAD_OK, 'success');
            Logger.log('File saved to', response);
          }).catch(err => {
            toast.show(STRINGS.TOAST.DOWNLOAD_ERROR, 'error');
            Logger.log('File save failure', err);
          });
        }
      });

      if (isMyMessage(msg, currentUserId) && msg.sendingStatus === 'succeeded') {
        sheetItems.push({
          icon: 'delete',
          title: STRINGS.GROUP_CHANNEL.DIALOG_MESSAGE_DELETE,
          onPress: () => confirmDelete(msg)
        });
      }

      const fileType = getFileType(msg.type || getFileExtension(msg.name));

      switch (fileType) {
        case 'image':
        case 'video':
        case 'audio':
          {
            response.onPress = () => {
              if (onPressImageMessage && fileType === 'image') {
                Logger.warn(DEPRECATION_WARNING.GROUP_CHANNEL.ON_PRESS_IMAGE_MESSAGE);
                onPressImageMessage(msg, getAvailableUriFromFileMessage(msg));
              }

              onPressMediaMessage === null || onPressMediaMessage === void 0 ? void 0 : onPressMediaMessage(msg, () => onDeleteMessage(msg), getAvailableUriFromFileMessage(msg));
            };

            break;
          }

        default:
          {
            response.onPress = () => SBUUtils.openURL(msg.url);

            break;
          }
      }
    }

    if (sheetItems.length > 0) {
      response.onLongPress = () => {
        openSheet({
          sheetItems,
          HeaderComponent: shouldRenderReaction(channel, features.reactionEnabled) ? _ref4 => {
            let {
              onClose
            } = _ref4;
            return /*#__PURE__*/React.createElement(ReactionAddons.BottomSheet, {
              message: msg,
              channel: channel,
              onClose: onClose
            });
          } : undefined
        });
      };
    }

    if (msg.sendingStatus === 'failed') {
      response.onLongPress = () => handleFailedMessage(msg);

      response.onPress = () => {
        onResendFailedMessage(msg).catch(() => toast.show(STRINGS.TOAST.RESEND_MSG_ERROR, 'error'));
      };
    }

    if (msg.sendingStatus === 'pending') {
      response.onLongPress = undefined;
      response.onPress = undefined;
    }

    return response;
  };
};

const styles = createStyleSheet({
  frozenBanner: {
    position: 'absolute',
    zIndex: 999,
    top: 8,
    left: 8,
    right: 8
  },
  frozenListPadding: {
    paddingBottom: 32
  },
  newMsgButton: {
    position: 'absolute',
    zIndex: 999,
    bottom: 10,
    alignSelf: 'center'
  },
  scrollButton: {
    position: 'absolute',
    zIndex: 998,
    bottom: 10,
    right: 16
  }
});
export default /*#__PURE__*/React.memo(GroupChannelMessageList);
