// @sendbird/uikit-utils/lib/module/sendbird/attrs.js
export let ApplicationAttributes;

(function (ApplicationAttributes) {
  ApplicationAttributes["reactions"] = "reactions";
  ApplicationAttributes["message_search_v3"] = "message_search_v3";
  ApplicationAttributes["enable_og_tag"] = "enable_og_tag";
  ApplicationAttributes["use_last_seen_at"] = "use_last_seen_at";
  ApplicationAttributes["enable_message_threading"] = "enable_message_threading";
  ApplicationAttributes["allow_message_update_from_sdk"] = "allow_message_update_from_sdk";
  ApplicationAttributes["allow_group_channel_create_from_sdk"] = "allow_group_channel_create_from_sdk";
  ApplicationAttributes["allow_migration_api"] = "allow_migration_api";
  ApplicationAttributes["allow_only_operator_sdk_to_update_group_channel_image"] = "allow_only_operator_sdk_to_update_group_channel_image";
  ApplicationAttributes["allow_user_update_from_sdk"] = "allow_user_update_from_sdk";
  ApplicationAttributes["allow_broadcast_channel"] = "allow_broadcast_channel";
  ApplicationAttributes["allow_sdk_file_message_urls"] = "allow_sdk_file_message_urls";
  ApplicationAttributes["allow_group_channel_leave_from_sdk"] = "allow_group_channel_leave_from_sdk";
  ApplicationAttributes["allow_user_list_from_sdk"] = "allow_user_list_from_sdk";
  ApplicationAttributes["allow_group_channel_invite_from_sdk"] = "allow_group_channel_invite_from_sdk";
  ApplicationAttributes["allow_only_operator_sdk_to_update_group_channel_name"] = "allow_only_operator_sdk_to_update_group_channel_name";
  ApplicationAttributes["allow_group_channel_update_from_sdk"] = "allow_group_channel_update_from_sdk";
  ApplicationAttributes["allow_super_group_channel"] = "allow_super_group_channel";
  ApplicationAttributes["allow_non_operators_to_mention_channel"] = "allow_non_operators_to_mention_channel";
  ApplicationAttributes["allow_file_messages_from_sdk"] = "allow_file_messages_from_sdk";
  ApplicationAttributes["allow_open_channel_create_from_sdk"] = "allow_open_channel_create_from_sdk";
  ApplicationAttributes["allow_anonymous_message_update"] = "allow_anonymous_message_update";
  ApplicationAttributes["allow_group_channel_delete_from_sdk"] = "allow_group_channel_delete_from_sdk";
  ApplicationAttributes["allow_operators_to_ban_operators"] = "allow_operators_to_ban_operators";
  ApplicationAttributes["allow_operators_to_edit_operators"] = "allow_operators_to_edit_operators";
})(ApplicationAttributes || (ApplicationAttributes = {}));

export let PremiumFeatures;

(function (PremiumFeatures) {
  PremiumFeatures["reactions"] = "reactions";
  PremiumFeatures["moderation_group"] = "moderation_group";
  PremiumFeatures["message_search_v3"] = "message_search_v3";
  PremiumFeatures["data_export"] = "data_export";
  PremiumFeatures["migration"] = "migration";
  PremiumFeatures["delivery_receipt"] = "delivery_receipt";
  PremiumFeatures["auto_thumbnail"] = "auto_thumbnail";
  PremiumFeatures["image_moderation"] = "image_moderation";
  PremiumFeatures["super_group_channel"] = "super_group_channel";
  PremiumFeatures["announcement"] = "announcement";
  PremiumFeatures["moderation_open"] = "moderation_open";
  PremiumFeatures["desk"] = "desk";
  PremiumFeatures["bot_interface"] = "bot_interface";
  PremiumFeatures["auto_trans"] = "auto_trans";
})(PremiumFeatures || (PremiumFeatures = {}));
