// @sendbird/uikit-react-native/lib/module/domain/groupChannel/module/moduleContext.js
import React, { createContext, useState } from 'react';
import { useChannelHandler } from '@sendbird/uikit-chat-hooks';
import { NOOP, isDifferentChannel, useUniqId } from '@sendbird/uikit-utils';
import ProviderLayout from '../../../components/ProviderLayout';
import { useLocalization, useSendbirdChat } from '../../../hooks/useContext';
export const GroupChannelContexts = {
  Fragment: /*#__PURE__*/createContext({
    headerTitle: '',
    channel: {},
    setMessageToEdit: NOOP
  }),
  TypingIndicator: /*#__PURE__*/createContext({
    typingUsers: []
  })
};
export const GroupChannelContextsProvider = _ref => {
  let {
    children,
    channel,
    enableTypingIndicator,
    keyboardAvoidOffset = 0
  } = _ref;
  if (!channel) throw new Error('GroupChannel is not provided to GroupChannelModule');
  const id = useUniqId('GroupChannelContextsProvider');
  const {
    STRINGS
  } = useLocalization();
  const {
    currentUser,
    sdk
  } = useSendbirdChat();
  const [typingUsers, setTypingUsers] = useState([]);
  const [messageToEdit, setMessageToEdit] = useState();
  useChannelHandler(sdk, `GroupChannelContextsProvider_${id}`, {
    onTypingStatusUpdated(eventChannel) {
      if (isDifferentChannel(channel, eventChannel)) return;
      if (!enableTypingIndicator) return;
      setTypingUsers(eventChannel.getTypingUsers());
    }

  });
  return /*#__PURE__*/React.createElement(ProviderLayout, null, /*#__PURE__*/React.createElement(GroupChannelContexts.Fragment.Provider, {
    value: {
      headerTitle: STRINGS.GROUP_CHANNEL.HEADER_TITLE((currentUser === null || currentUser === void 0 ? void 0 : currentUser.userId) ?? '', channel),
      channel,
      messageToEdit,
      setMessageToEdit,
      keyboardAvoidOffset
    }
  }, /*#__PURE__*/React.createElement(GroupChannelContexts.TypingIndicator.Provider, {
    value: {
      typingUsers
    }
  }, children)));
};
