// @sendbird/uikit-utils/lib/commonjs/sendbird/channel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.confirmAndMarkAsRead = exports.confirmAndMarkAsDelivered = void 0;
exports.getChannelUniqId = getChannelUniqId;
exports.getGroupChannelChatAvailableState = void 0;
exports.getGroupChannels = getGroupChannels;
exports.getMembersExcludeMe = getMembersExcludeMe;
exports.getOpenChannels = getOpenChannels;
exports.isDefaultCoverImage = isDefaultCoverImage;
exports.isDifferentChannel = isDifferentChannel;

var _bufferedRequest = require("../shared/bufferedRequest");

/**
 * Diff utils for channel
 * @param {SendbirdBaseChannel} [a]
 * @param {SendbirdBaseChannel} [b]
 * @returns {boolean}
 * */
function isDifferentChannel(a, b) {
  if (!a || !b) return true;
  return a.url !== b.url;
}

const getGroupChannelChatAvailableState = channel => {
  const frozen = channel.isFrozen && channel.myRole !== 'operator';
  const muted = channel.myMutedState === 'muted';
  const disabled = frozen || muted;
  return {
    disabled,
    frozen,
    muted
  };
};

exports.getGroupChannelChatAvailableState = getGroupChannelChatAvailableState;

const confirmAndMarkAsRead = async channels => {
  channels.filter(it => it.isGroupChannel() && it.unreadMessageCount > 0).forEach(it => _bufferedRequest.BufferedRequest.markAsRead.push(() => it.markAsRead(), it.url));
};

exports.confirmAndMarkAsRead = confirmAndMarkAsRead;

const confirmAndMarkAsDelivered = async channels => {
  channels.filter(it => it.isGroupChannel() && it.unreadMessageCount > 0).forEach(it => _bufferedRequest.BufferedRequest.markAsDelivered.push(() => it.markAsDelivered(), it.url));
};

exports.confirmAndMarkAsDelivered = confirmAndMarkAsDelivered;

function isDefaultCoverImage(coverUrl) {
  return coverUrl === '' || coverUrl.startsWith('https://static.sendbird.com/sample/cover');
}

function getMembersExcludeMe(channel, currentUserId) {
  return channel.members.filter(m => m.userId !== currentUserId);
}

function getGroupChannels(channels) {
  return channels.filter(c => c.isGroupChannel());
}

function getOpenChannels(channels) {
  return channels.filter(c => c.isOpenChannel());
}

function getChannelUniqId(channel) {
  return channel.url;
}
