// @sendbird/uikit-react-native/lib/commonjs/domain/groupChannel/component/GroupChannelInput/EditInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _message = require("@sendbird/chat/message");

var _uikitReactNativeFoundation = require("@sendbird/uikit-react-native-foundation");

var _useContext = require("../../../../hooks/useContext");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const EditInput = /*#__PURE__*/(0, _react.forwardRef)(function EditInput(_ref, ref) {
  let {
    text,
    onChangeText,
    messageToEdit,
    setMessageToEdit,
    onUpdateUserMessage,
    onSelectionChange,
    disabled,
    autoFocus,
    mentionedUsers
  } = _ref;
  const {
    mentionManager
  } = (0, _useContext.useSendbirdChat)();
  const {
    STRINGS
  } = (0, _useContext.useLocalization)();
  const toast = (0, _uikitReactNativeFoundation.useToast)();

  const onPressCancel = () => {
    setMessageToEdit();
    onChangeText('');
  };

  const onPressSave = () => {
    if (messageToEdit.isUserMessage()) {
      const mention = {
        userIds: mentionedUsers.map(it => it.user.userId),
        messageTemplate: mentionManager.textToMentionedMessageTemplate(text, mentionedUsers),
        type: _message.MentionType.USERS
      };
      onUpdateUserMessage(text, messageToEdit, mention).catch(() => toast.show(STRINGS.TOAST.UPDATE_MSG_ERROR, 'error'));
    }

    setMessageToEdit();
    onChangeText('');
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.editInputContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.inputWrapper
  }, /*#__PURE__*/_react.default.createElement(_uikitReactNativeFoundation.TextInput, {
    ref: ref,
    multiline: true,
    disableFullscreenUI: true,
    editable: !disabled,
    autoFocus: autoFocus,
    onChangeText: onChangeText,
    style: styles.input,
    placeholder: STRINGS.GROUP_CHANNEL.INPUT_PLACEHOLDER_ACTIVE,
    onSelectionChange: onSelectionChange
  }, mentionManager.textToMentionedComponents(text, mentionedUsers))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      marginTop: 8,
      flexDirection: 'row'
    }
  }, /*#__PURE__*/_react.default.createElement(_uikitReactNativeFoundation.Button, {
    variant: 'text',
    onPress: onPressCancel
  }, STRINGS.GROUP_CHANNEL.INPUT_EDIT_CANCEL), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.space
  }), /*#__PURE__*/_react.default.createElement(_uikitReactNativeFoundation.Button, {
    variant: 'contained',
    onPress: onPressSave
  }, STRINGS.GROUP_CHANNEL.INPUT_EDIT_OK)));
});
const styles = (0, _uikitReactNativeFoundation.createStyleSheet)({
  editInputContainer: {
    paddingVertical: 10,
    paddingHorizontal: 16,
    flexDirection: 'column',
    alignItems: 'center'
  },
  input: {
    flex: 1,
    marginRight: 4,
    minHeight: 36,
    maxHeight: 36 * _reactNative.Platform.select({
      ios: 2.5,
      default: 2
    }),
    borderRadius: 20
  },
  inputWrapper: {
    flexDirection: 'row'
  },
  space: {
    flex: 1
  }
});
var _default = EditInput;
exports.default = _default;
